### ISSUE TITLE
Python 2 and 3 paths are mixed up

### ISSUE THREAD

**Issue by Ekleog** [NixOS/nixpkgs#49396]
Reproducer: `nix-shell '<nixpkgs>' -A notmuch --run "python3"` (with `python3` in the path). Notice that there is a traceback mixing up python 2 and python 3 libraries.

`notmuch` requires `python2` libraries (in the current nixpkgs setup). But its test framework requires `python3`, hence the issue.

I think we can do better than crash horribly when running python 3 with python 2 libraries in the path. Idea: don't set directly `PYTHONPATH`, but set `NIX_PYTHON2PATH` and `NIX_PYTHON3PATH`, and use a wrapper around `python2` and `python3` to add `NIX_PYTHON2PATH` and `NIX_PYTHON3PATH` at the end of `PYTHONPATH` (so that users can override the path if they want).

What do you think about this?

cc @FRidh

**Comment by Ekleog** [comment]
Indeed, and following through your links I also came upon https://github.com/NixOS/nixpkgs/pull/792 that was another attempt at implementing the same idea, that got closed under the idea that there were better ways to do this…

Actually, I don't see a reason not to implement it, especially with the fact someone is asking for it every year or so. It's not (IMO) because there are better known solutions that we should allow ourselves to crash randomly when the wrong way is used? :)

**Comment by Ekleog** [comment]
@FRidh 

> First of all, nix-shell is for development of a derivation. If you just want to run something, use nix run. Problem solved.

I don't want to run something, I want to hack on `notmuch`. Problem not solved, unless `nix run` has an undocumented equivalent to `nix-shell -A`.

> Second, separating Python 2 and 3 solves just a fraction of the issues. One can mix also various Python 2 environments, which will also cause trouble, though more likely silently.

I didn't know mixing two different versions of python2 could cause trouble. However, #792 does handle the case of different python2 versions.

> And for this we have a solution: python.withPackages.

This is not a solution, this is a workaround. The current behavior is completely confusing, even if it is documented, and the existence of `python.withPackages` doesn't prevent it. If some big warning message pointing to `python.withPackages` was displayed upon entering a shell with a python package I'd agree with you, but there is none, and as you can note from my example above I did not even have python packages in my shell command line.

Basically, what I read in your message (hoping I don't read too much in it) is “well, it's documented that we do nonsense when using `nix-shell` with python packages, so wontfix”. My point is that we should either stop doing nonsense (#792), or at least fail with a big error message.

**Comment by Ekleog** [comment]
This sounds like the same problems C libraries have (once limited to a single python version), and C libraries work relatively well in a `nix-shell`… much better than the current state of Python libraries at least.

Anyway, as you appear to be firmly opposed to the idea, can we please reopen this on the theme of “opening a `nix-shell` with python libraries should display a big warning”? It would at least avoid the recurrent problems, and people not even noticing they are entering a shell with python libraries (like me, and I was very “wtf‽” when I had this stack trace mixing python2 and python3 code without having added anything python-related to my environment or shell -- or so I thought)…

**Comment by siraben** [comment]
Any updates on this?
